apm-agent-plugins/apm-es-restclient-plugin/src/main/java/co/elastic/apm/es/restclient/ElasticsearchRestClientInstrumentation.java

```diff
@@ -0,0 +1,138 @@
+ * Copyright (C) 2018 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.es.restclient;
+
+import co.elastic.apm.bci.ElasticApmInstrumentation;
+import co.elastic.apm.bci.VisibleForAdvice;
+import co.elastic.apm.impl.transaction.Span;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.elasticsearch.client.Request;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.ResponseException;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+/**
+ * Instrumentation for Elasticsearch RestClient, currently supporting only synchronized queries.
+ * All sync operations go through org.elasticsearch.client.RestClient#performRequest(org.elasticsearch.client.Request)
+ */
+public class ElasticsearchRestClientInstrumentation extends ElasticApmInstrumentation {
+    @VisibleForAdvice
+    public static final String SEARCH_QUERY_PATH_SUFFIX = "_search";
+
+    @VisibleForAdvice
+    static final String SPAN_TYPE = "db.elasticsearch.request";
+    @VisibleForAdvice
+    static final String DB_CONTEXT_TYPE = "elasticsearch";
+
+    @VisibleForAdvice
+    static final String ELASTICSEARCH_NODE_KEY = "Elasticsearch-node";
+    @VisibleForAdvice
+    static final String QUERY_STATUS_CODE_KEY = "Query-status-code";
+    @VisibleForAdvice
+    static final String ERROR_REASON_KEY = "Error-reason";
+
+    @Advice.OnMethodEnter
+    private static void onBeforeExecute(@Advice.Argument(0) Request request,
+                                        @Advice.Local("span") Span span) {
+        if (tracer == null || tracer.getActive() == null) {
+            return;
+        }
+        span = tracer.getActive().createSpan()
+            .withType(SPAN_TYPE)
+            .appendToName("Elasticsearch: ").appendToName(request.getMethod()).appendToName(" ").appendToName(request.getEndpoint())
+            .activate();
+
+        if (request.getEndpoint().endsWith(SEARCH_QUERY_PATH_SUFFIX)) {
```

Comment:
add `&& span.isSampled()` to avoid overhead on non sampled spans

Reply:
Actually, shouldn't this check be:
```
        if (tracer == null || tracer.getActive() == null || !tracer.getActive().isSampled()) {
            return;
        }
``` 
at the beginning of this method?

Reply:
Good point. It should probably be both. In case the span limit (default 500 per transaction) is exceeded, the parent might be sampled but the span not.